dol/src/lib.rs

```diff
@@ -107,16 +109,36 @@ impl Dol {
         })
     }
 
+    pub fn section_data(&self, section: &DolSection) -> &[u8] {
+        self.virtual_data_at(section.target, section.size).unwrap()
+    }
+
+    /// Returns a slice of DOL data. Does not support bss.
+    pub fn virtual_data_at(&self, virtual_addr: u32, read_len: u32) -> Result<&[u8]> {
+        let offset = (virtual_addr - self.memory_offset) as usize;
+        if offset + (read_len as usize) < self.memory.len() {
+            Ok(&self.memory[offset..offset + (read_len as usize)])
+        } else {
+            Err(Error::OutOfBound(virtual_addr + read_len))
+        }
+    }
+
     /// Reads bytes into a destination buffer given a virtual address.
-    pub fn virtual_read(&self, dest: &mut [u8], virtual_addr: u32) {
+    pub fn virtual_read(&self, data: &mut [u8], virtual_addr: u32) -> Result<()> {
         let offset = (virtual_addr - self.memory_offset) as usize;
```

Comment:
If `virtual_addr` is smaller than `self.memory_offset`, this will panic on debug mode (integer underflow checks), so this also needs a bounds check.